libraries/botframework-connector/src/auth/serviceClientCredentialsFactory.ts

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { ServiceClientCredentials } from '@azure/ms-rest-js';
+
+/**
+ * The ServiceClientCredentialsFactory abstract class that allows Bots to provide their own
+ * ServiceClientCredentials for bot to bot channel or skill bot to parent bot calls.
+ */
+export abstract class ServiceClientCredentialsFactory {
+    /**
+     * Validates an app ID.
+     *
+     * @param appId The app ID to validate.
+     * @returns {Promise<boolean>} The result is true if `appId` is valid for the controller; otherwise, false.
+     */
+    abstract isValidAppId(appId: string): Promise<boolean>;
+
+    /**
+     * Checks whether bot authentication is disabled.
+     * @returns {Promise<boolean>} If bot authentication is disabled, the result is true; otherwise, false.
+     */
+    abstract isAuthenticationDisabled(): Promise<boolean>;
+
+    /**
+     * A factory method for creating ServiceClientCredentials.
+     *
+     * @param appId The appId.
+     * @param audience The audience.
+     * @param loginEndpoint The login url.
+     * @param validateAuthority The validate authority vale to use.
+     * @returns {Promise<ServiceClientCredentials>} A [ServiceClientCredentials](xref:botframework-connector.ServiceClientCredentials).
+     */
+    abstract createCredentials(
+        appId: string,
+        audience: string,
+        loginEndpoint: string,
+        validateAuthority: boolean
+    ): Promise<ServiceClientCredentials>;
```

Comment:
Any chance of this needing to return `null` or `undefined`?

Reply:
Good question! After checking the .NET implementation, it's only called after a request is verified, so there are no auth errors.

The internal .NET classes (yet-to-be-ported) also require a ServiceClientCredentials object, and in our case the ConnectorClient contract we follow (because we don't have an abstract class or interface for the ConnectorClient) requires a ServiceClientCredentials object to be instantiated.

I think it's safe to say that this method will have an extremely low chance of needing to return `null` or `undefined`.